File: PA6.java
class PA6{
    
    public static int search(int arr[], int target){
        int low= 0;
        int high =1;
       
        while(low<=high){

            if(target<=arr[high]){
                int mid=0;
                while(low<=high){
                    mid = low+(high-low)/2;
                    if(arr[mid]== target){
                        return mid;
                    }
                    if(arr[mid]<target){
                        low= mid+1;
                    }
                    else{
                        high= mid-1;
                    }
                }
                return -1;
            }
            else{
                low= high;
                //if(2*high > arr.length-1){
                  //  high = arr.length-1;
               // }
                //else{
                    high= 2*high;
               // }
                
            }
        }

        return -1;
    }
    public static void main(String args[]) {
        int[] arr = {1,2,3,4,5,6,7,8,9,14,17,21,28,29,33,36,39,58,63,66,72};
        int target= 7;
       System.out.println("Target "+target+" found at index "+search(arr, target)); 
    }
}

File: PA_6.py
#Leetcode 702 : Search in a Sorted Array of Unknown Size
#Time Complexity : O(logn) [for both, binarySearch() and getRange()]
#Space Complexity : O(1)

# """
# This is ArrayReader's API interface.
# You should not implement it, or speculate about its implementation
# """
#class ArrayReader(object):
#    def get(self, index):
#        """
#        :type index: int
#        :rtype int
#        """


#This is the leetcode version of the question. Instead of using the reader, we can simply do arr[index] to get the value.
class Solution(object):
    def search(self, reader, target):
        """
        :type reader: ArrayReader
        :type target: int
        :rtype: int
        """
        low,high = self.getRange(reader,target)
        return self.binarySearch(low,high,target,reader)
    
    def binarySearch(self,low,high,target,reader):
        while low <= high:
            mid = (low + high) // 2
            if reader.get(mid) == target:
                return mid
            elif reader.get(mid) < target:
                low = mid + 1
            else:
                high = mid - 1
        return -1
    
    def getRange(self,reader,target):
        low = 0
        high = 1
        while reader.get(high) < target:
            low = high
            high = 2 * high
        return [low,high]